layers: Allow using uncached buffer for Debug Printf

dorian-apanel-intel opened the pull request:
Adds option to force using AMD_DEVICE_COHERENT_MEMORY for debug printf buffer to print messages even if VK_ERROR_DEVICE_LOST is encountered.

Forcing extension and device feature if not enabled by application.

Added workaround for atomic operations in
uncached memory being in cache anyway.

Added workaround to failing MapMemory after DEVICE_LOST (occurs on AMD): When using uncached buffer, do not unmap buffer until messages are analyzed.

This PR attempts to implement #6101 Tested on AMD and Intel (extension not public yet) using this modified vkcube: https://github.com/dorian-apanel-intel/Vulkan-Tools/commit/e0226930a7958bb4964428691dd1376c3447bca3

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/Vulkan-ValidationLayers?pullRequest=6128) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/Vulkan-ValidationLayers?pullRequest=6128) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/Vulkan-ValidationLayers?pullRequest=6128) it.</sub>

ci-tester-lunarg commented:
Author dorian-apanel-intel not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author dorian-apanel-intel not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author dorian-apanel-intel not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author dorian-apanel-intel not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author dorian-apanel-intel not on autobuild list. Waiting for curator authorization before starting CI build.

jeremyg-lunarg commented on layers/gpu_validation/debug_printf.cpp:
```diff
@@ -680,11 +819,16 @@ void DebugPrintf::AllocateDebugPrintfResources(const VkCommandBuffer cmd_buffer,
+            vmaUnmapMemory(vmaAllocator, output_block.allocation);
+            output_block.data = nullptr;
+        } else {
+            // Mapping may fail after DEVICE_LOST. Keep it mapped for now.
```
I'd prefer just the comment above the if block rather than this empty else block.

dorian-apanel-intel replied:
Done

jeremyg-lunarg commented on layers/gpu_validation/debug_printf.cpp:
```diff
@@ -138,6 +228,42 @@ void DebugPrintf::PreCallRecordCreateShaderModule(VkDevice device, const VkShade
+    bool device_lost = (result == VK_ERROR_DEVICE_LOST);
+
+    if (aborted) return;
+    if (!((result == VK_SUCCESS) || (device_lost && use_uncached_buffer))) return;
```
Should this be `...(device_lost && !use_uncached_bufffer))) return;`

It seems to me that if we're using uncached buffers to get data after a device lost we'd want to process them here. Or am I missing something?

dorian-apanel-intel replied:
Maybe you missed that negation at start `!`
I will try to write it cleaner.

jeremyg-lunarg replied:
yeah I did miss that.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 5177.

Tony-LunarG reviewed (commented):
Has this been tested?  Is there a test that can go with this?

juan-lunarg reviewed (commented):
I don't see any testing being added. Is the plan to add them later?

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 12676 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 12676 passed.

ci-tester-lunarg commented:
Author dorian-apanel-intel not on autobuild list. Waiting for curator authorization before starting CI build.

dorian-apanel-intel commented:
@juan-lunarg @Tony-LunarG 
> I don't see any testing being added. Is the plan to add them later?
> Has this been tested? Is there a test that can go with this?

I've only tested it manually so far.
Would it be enough to copy one test from `tests\negative\debug_printf.cpp` and modify it to use uncached buffer and hang?

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 12692 running.

Tony-LunarG commented:
> @juan-lunarg @Tony-LunarG
> 
> > I don't see any testing being added. Is the plan to add them later?
> > Has this been tested? Is there a test that can go with this?
> 
> I've only tested it manually so far. Would it be enough to copy one test from `tests\negative\debug_printf.cpp` and modify it to use uncached buffer and hang?

I think so, as long as the test gets a DEVICE_LOST and then successfully parses a debug printf

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 12692 passed.

juan-lunarg commented:
>  Would it be enough to copy one test from tests\negative\debug_printf.cpp and modify it to use uncached buffer and hang?

This sounds like it would cause problems for Jenkins CI (Although if we switched to ctest maybe not... since the test would be in it's own process).

Because currently CI invokes the test executable and tries running all tests.

ci-tester-lunarg commented:
Author dorian-apanel-intel not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author dorian-apanel-intel not on autobuild list. Waiting for curator authorization before starting CI build.

dorian-apanel-intel commented:
@Tony-LunarG @juan-lunarg @jeremyg-lunarg
* I've added unit test, had to pass a value using push constants, because hardcoded infinite loop was not hanging on AMD.
* I modified layer_options.cpp file to move option from test to layer. It is not perfect, but after looking at #5895, I think it will need rewrite soon anyway.
* After updating AMD drivers, Atomics WA is no longer needed there, but leaving it for Intel for now.
* NVidia does not support required extension yet.
* Force push was required due to printf test file rename.

I plan to add also unit test for Page Fault, but preferably in another pull request.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 7794.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 12728 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 12728 failed.

juan-lunarg commented:
Galaxy S22 testing failed

```
[ RUN      ] NegativeDebugPrintf.UncachedBuffer
jni/../../tests/framework/error_monitor.cpp:200: Failure
Failed
Validation Error: [ UNASSIGNED-DEBUG-PRINTF ] Object 0: handle = 0x6c1d064220, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x92394c89 | Setup Error. Detail: (Unable to allocate device memory.  Device could become unstable.)
jni/../../tests/framework/error_monitor.cpp:241: Failure
Failed
Did not receive expected error 'Here are three float values 1.000000, 3.141500, 0.000000'
[  FAILED  ] NegativeDebugPrintf.UncachedBuffer (5836 ms)
```

juan-lunarg reviewed (commented):
LGTM. Thanks for the test 👍🏾 

I'll defer to @Tony-LunarG and @jeremyg-lunarg for the actual review.

Tony-LunarG commented:
> Galaxy S22 testing failed
> 
> ```
> [ RUN      ] NegativeDebugPrintf.UncachedBuffer
> jni/../../tests/framework/error_monitor.cpp:200: Failure
> Failed
> Validation Error: [ UNASSIGNED-DEBUG-PRINTF ] Object 0: handle = 0x6c1d064220, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x92394c89 | Setup Error. Detail: (Unable to allocate device memory.  Device could become unstable.)
> jni/../../tests/framework/error_monitor.cpp:241: Failure
> Failed
> Did not receive expected error 'Here are three float values 1.000000, 3.141500, 0.000000'
> [  FAILED  ] NegativeDebugPrintf.UncachedBuffer (5836 ms)
> ```

I don't think there's much the test can do about that.  I'll just add the new test to the S22 blacklist.

Tony-LunarG commented on layers/gpu_validation/debug_printf.cpp:
```diff
@@ -308,10 +441,14 @@ void DebugPrintf::AnalyzeAndGenerateMessages(VkCommandBuffer command_buffer, VkQ
 
     uint32_t index = spvtools::kDebugOutputDataOffset;
-    while (debug_output_buffer[index]) {
+    while ((index < output_buffer_size) && debug_output_buffer[index]) {
```
I don't think this loop will work if the output buffer has one message but isn't full.  To see this, use uncached buffer memory in the test NegativeDebugPrintf.BasicUsage and run it. This loop doesn't stop when it's supposed to.

dorian-apanel-intel replied:
Could you check in which vkQueueSubmit/AnalyzeAndGenerateMessages you observe it? BasicUsage does it 15 times.

Here is how I understand it:

In my first one, `debug_output_buffer == 0x0000028b8e520000`
Memory of that buffer is:

```
0x0000028B8E520000  00 00 00 00  MBZ
0x0000028B8E520004  0a 00 00 00  expect (size of all messages, written by atomic in shader)
0x0000028B8E520008  0a 00 00 00  DPFOutputRecord.size (size in dwords of this message)
0x0000028B8E52000C  00 00 00 00  DPFOutputRecord.shader_id
0x0000028B8E520010  70 00 00 00  DPFOutputRecord.instruction_position
0x0000028B8E520014  00 00 00 00  DPFOutputRecord.stage
0x0000028B8E520018  00 00 00 00  DPFOutputRecord.stage_word_1
0x0000028B8E52001C  00 00 00 00  DPFOutputRecord.stage_word_2
0x0000028B8E520020  00 00 00 00  DPFOutputRecord.stage_word_3
0x0000028B8E520024  29 00 00 00  DPFOutputRecord.format_string_id ("Here are two float values %f, %f")
0x0000028B8E520028  00 00 80 3f  DPFOutputRecord.values (float 1.0)
0x0000028B8E52002C  56 0e 49 40  DPFOutputRecord.values (float 3.14150000)
0x0000028B8E520030  00 00 00 00  DPFOutputRecord.size  (no more messages)
0x0000028B8E520034  00 00 00 00  DPFOutputRecord.
0x0000028B8E520038  00 00 00 00  DPFOutputRecord.
```

Index starts at 0x2. At the end of first loop iteration:
`index += debug_record->size; // 0x2 + 0xa = 0xc`
At second loop beginning 
`debug_output_buffer[index]` == 0x0000028b8e520030 {0x0}
So the loop ends.

`(index < output_buffer_size)` means "while we are still in the buffer"
`&& debug_output_buffer[index])` means "while next message is not empty"

Tony-LunarG replied:
I don't have access to the system I was using to test this today - I'll get back to it tomorrow.  The system I do have is getting VK_TIMEOUT instead of VK_ERROR_DEVICE_LOST, and I had to add that to the test's expected results. You should add that to your test.

Tony-LunarG replied:
I see the problem.  Because debug_output_buffer[1] comes back as 0, the memset at the end of message processing doesn't clear the buffer:
```
memset(debug_output_buffer, 0, 4 * (debug_output_buffer[spvtools::kDebugOutputSizeOffset] + spvtools::kDebugOutputDataOffset));
```
And the memset clears 4 * (0 + 2) bytes which doesn't cover all of the records.  Then on the next printf, the loop depends on ```debug_output_buffer[index]``` being zero when the loop is supposed to stop, but previous printfs may have left values that didn't get memset to 0.  I suppose one answer would be to memset the whole buffer.

But I'm seeing another problem.  If I modify your test to do two debug printfs:
```
          if (gl_VertexIndex == 0) {
                debugPrintfEXT("Here are three float values %f, %f, %f", 1.0, myfloat, gl_Position.x);
                debugPrintfEXT("Here's another debug printf");
                float x = constants.x[0];
                while(x > -1.f) { // infinite loop
                    x += 0.000001f;
                }
                debugPrintfEXT("Here is a value that should not be printed %f", x);
            }
```
I only get the second printf back, as if the first one was overwritten in the shader.  Could that be a result of the atomic operation problem and debug_output_buffer[1] not getting written?
(This is all on an AMD RX6600 with driver 23.7.1)

dorian-apanel-intel replied:
Thanks for finding that memset issue, I've added whole buffer clear in case of uncached buffer.

The only second printf working I cannot reproduce on AMD RX 5700 XT with the same driver you have. Maybe I will ask AMD devs for feedback.

Tony-LunarG replied:
My working theory was that the shader instrumentation was using data_buffer[1] as an offset to write the next debug printf result, so if it's zero, all printfs would overwrite each other.  But it sounds like you've tested a case where [1] is not written, but you still get 2 printf results?

Tobski replied:
How is the problematic AtomicAdd performed? I don't know where to look for the code to try to figure this out.

Tobski replied:
I may have uncovered some driver weirdness here so I've started a discussion internally to see whether we've messed something up here, I'll let you know if we confirm a bug.

Tony-LunarG replied:
The atomic add is performed by the instrumentation added to the shader via spirv-tools code, which I'm not very familiar with.

jeremyg-lunarg replied:
https://github.com/KhronosGroup/SPIRV-Tools/blob/main/source/opt/instrument_pass.cpp#L823

Tobski replied:
Ok so after a bit of digging, this extension was enabled before we had the MALL cache in hardware, and nobody ever went back and changed the interface to bypass MALL, so there IS caching enabled on this memory type for newer hardware. Whoops. I've filed an internal bug to fix this and to add a more comprehensive test to make sure this actually works on the off chance we make another big change to the caching hierarchy...

You can go look at the public driver source if for some reason you want to see where this is happening 🙃

I'll let you know when it's fixed and subsequently released.

dorian-apanel-intel replied:
@Tobski I assume the fix is released now?

Tobski replied:
A fix has been implemented but I'm not sure on release status, I'll take a look.

Tony-LunarG reviewed (changes requested):
The message processing loop won't exit correctly

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 9060.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 12764 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 12764 passed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 10076.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 12781 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 12781 failed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 10420.

lunarpapillo commented:
Last LunarG CI failure was due to a system issue.  I've restarted.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 12787 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 12787 passed.

Tobski commented on layers/gpu_validation/debug_printf.cpp:
```diff
@@ -672,6 +821,9 @@ void DebugPrintf::AllocateDebugPrintfResources(const VkCommandBuffer cmd_buffer,
     VmaAllocationCreateInfo alloc_info = {};
     alloc_info.requiredFlags = VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;
+    if (use_uncached_buffer) {
+        alloc_info.requiredFlags |= VK_MEMORY_PROPERTY_DEVICE_COHERENT_BIT_AMD | VK_MEMORY_PROPERTY_DEVICE_UNCACHED_BIT_AMD;
```
You probably don't want to require UNCACHED here, since _theoretically_ in future you might get COHERENT _and_ CACHED memory. Uncached is meant to be informative, not something you generally rely upon unless you're doing some very specific performance tuning.

dorian-apanel-intel replied:
I don't quite understand. Here I want to get device uncached to read it after device is lost.

spencer-lunarg commented:
this has got stale - closing, but tracking here (https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/9211) so can reference this in future
